java/src/com/ibm/streamsx/topology/kafka/KafkaStreams.java

```diff
@@ -0,0 +1,284 @@
+       /*
+         File tmpDir = mkTmpDir
+         File etcDir = mkDir(tmpDir, "etc");
+         File props = mkFile(tmpDir, PROPS_FILE);
+         props.createNewFile();
+         
+        te.addFileDependency(props.getAbsolutePath());
+        
+        when will it get removed?  if instead Topology makes the copy...???
+         */
+    }
+
+    /**
+     * Convert an {@link SPLStream} with schema {@link KafkaSchemas.KAFKA}
+     * to a TStream&lt;{@link KafkaMessageImpl}>.
+     * The returned stream will contain a {@code KafkaTuple} tuple for
+     * each tuple {@code T} on {@code stream}.
+     * A runtime error will occur if the schema of {@code stream} doesn't
+     * have the attributes defined by {@code KafkaSchemas.KAFKA}.
+     * @param stream Stream to be converted to a TStream&lt;KafkaTuple>.
+     * @return Stream of {@code KafkaTuple} tuples from {@code stream}.
+     */
+    private static TStream<KafkaMessage> toKafkaTupleStream(SPLStream stream) {
+
+        return stream.convert(new Function<Tuple, KafkaMessage>() {
+            private static final long serialVersionUID = 1L;
+
+            @Override
+            public KafkaMessage apply(Tuple tuple) {
+                return new KafkaMessageImpl(fromSplValue(tuple.getString("message")),
+                                    fromSplValue(tuple.getString("key")));
+            }
+            
+            private String fromSplValue(String s) {
+                // SPL doesn't allow null value.  For our use,
+                // assume an empty string meant null.
+                return (s==null || s.isEmpty()) ? null : s;
+            }
+        }, KafkaMessage.class);
+    }
+
+    /**
+     * Send a stream of messages to a Kafka Cluster.
+     * This is an adapter to a Kafka KafkaProducer.
+     * <p>
+     * N.B., A topology that includes this will not support
+     * {@code StreamsContext.EMBEDDED}.
+     *  
+     * @param kafkaProducerProperties see the Kafka documentation.
+     *          Minimal configuration typically includes:
+     * <ul>
+     * <li><code>metadata.broker.list</code></li>
+     * <li><code>serializer.class</code></li>
+     * <li><code>request.required.acks</code></li>
+     * </ul>
+     * @param topic optional Kafka topic to publish to.
+     *          Whens specified, a {@link KafkaMessageImpl#getKafkaTopic()} value is ignored.
+     *          If null, {@code toKafkaTupleFunc} must generate tuples that have a topic.
+     * @param stream the stream to publish to Kafka
+     * @param toKafkaMessageFunc function to generate a {@code KafkaTuple} from a tuple {@code T}.
+     *          If {@code topic} is null, the generated {@code KafkaTuple} <b>must</b> contain a {@code topic}.
+     */
+    public static <T> void sink(TopologyElement te,
+            Properties kafkaProducerConfig,
+            final String topic,
+            TStream<T> stream,
+            final Function<T,? extends KafkaMessage> toKafkaMessageFunc)
```

Comment:
There's an asymmetry in that TStream<KafkaMessage> is produced but the consumer does not take TStream<KafkaMessage>. Instead it takes any TStream<T> but introduces a new way to do tuple transformation.

An existing pattern would be to accept TStream<T extends KafkaMessage> and expect that TStream.transform() would be used for conversion. This would require the interface for KafkaMessage

Or maybe follow the pattern of JSON, where classes can implement an interface to convert to KafkaMessage (KafkaMessageable (really bad name)).

Not saying what is done is wrong, but just trying to think why a different approach was taken, and will we end up with many similar methods for all different types of output/input classes, thus transformation of tuples from one type to another is no longer a single building block?

Not sure